src/diffusers/models/transformers/sana_transformer_2d.py

```diff
@@ -0,0 +1,702 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Dict, Optional, Union
+
+import torch
+from torch import nn
+
+from ...configuration_utils import ConfigMixin, register_to_config
+from ...utils import is_torch_version, logging
+from ...utils.torch_utils import maybe_allow_in_graph
+from ..attention_processor import (
+    Attention,
+    AttentionProcessor,
+    AttnProcessor2_0,
+    FusedAttnProcessor2_0,
+    SanaLinearAttnProcessor2_0,
+)
+from ..embeddings import PatchEmbed, PixArtAlphaTextProjection, SinusoidalPositionalEmbedding
+from ..modeling_outputs import Transformer2DModelOutput
+from ..modeling_utils import ModelMixin
+from ..normalization import AdaLayerNormSingle, RMSNorm
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+def _chunked_feed_forward(ff: nn.Module, hidden_states: torch.Tensor, chunk_dim: int, chunk_size: int, HW: tuple=None):
+    # "feed_forward_chunk_size" can be used to save memory
+    if hidden_states.shape[chunk_dim] % chunk_size != 0:
+        raise ValueError(
+            f"`hidden_states` dimension to be chunked: {hidden_states.shape[chunk_dim]} has to be divisible by chunk size: {chunk_size}. Make sure to set an appropriate `chunk_size` when calling `unet.enable_forward_chunking`."
+        )
+
+    num_chunks = hidden_states.shape[chunk_dim] // chunk_size
+    ff_output = torch.cat(
+        [ff(hid_slice, HW) for hid_slice in hidden_states.chunk(num_chunks, dim=chunk_dim)],
+        dim=chunk_dim,
+    )
+    return ff_output
+
+
+@maybe_allow_in_graph
+class RMSNormScaled(RMSNorm):
+    def __init__(self, dim, eps: float, elementwise_affine: bool = True, scale_factor: float = 1.0):
+        super().__init__(dim, eps, elementwise_affine)
+        self.weight = nn.Parameter(torch.ones(dim) * scale_factor)
+
+
+# Modified from diffusers.models.autoencoders.autoencoder_dc.GLUMBConv
+@maybe_allow_in_graph
+class SanaGLUMBConv(nn.Module):
```

Comment:
Hmm... so we implemented this in the autoencoder file. But since it is also re-used here, I'm not sure where to best place this layer so it can be shared. @yiyixuxu what do you think should be done?

Reply:
I'm afraid the GLUMBConv here is not the same with the one in autoencoder file. Specifically, the operation channels in these two operators are not the same. and the height width dimensions either.

```python
def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
    residual = hidden_states

    hidden_states = self.conv_inverted(hidden_states)
    hidden_states = self.nonlinearity(hidden_states)

    hidden_states = self.conv_depth(hidden_states)
    hidden_states, gate = torch.chunk(hidden_states, 2, dim=1)
    hidden_states = hidden_states * self.nonlinearity(gate)

    hidden_states = self.conv_point(hidden_states)
    # move channel to the last dimension so we apply RMSnorm across channel dimension
    hidden_states = self.norm(hidden_states.movedim(1, -1)).movedim(-1, 1)

    return hidden_states + residual
```
vs
```python
def forward(self, hidden_states: torch.Tensor, HW: Optional[tuple[int]] = None) -> torch.Tensor:
    B, N, C = hidden_states.shape
    if HW is None:
        H = W = int(N**0.5)
    else:
        H, W = HW

    hidden_states = hidden_states.reshape(B, H, W, C).permute(0, 3, 1, 2)

    hidden_states = self.conv_inverted(hidden_states)
    hidden_states = self.nonlinearity(hidden_states)

    hidden_states = self.conv_depth(hidden_states)
    hidden_states, gate = torch.chunk(hidden_states, 2, dim=1)
    hidden_states = hidden_states * self.nonlinearity(gate)

    hidden_states = self.conv_point(hidden_states)
    hidden_states = hidden_states.reshape(B, C, N).permute(0, 2, 1)

    return hidden_states
```

Reply:
I see. I think the two could be combined and I'll take this up for refactor. If it doesn't work out, it's okay to have a second class